### PR TITLE
Caching Implementation

### PR DESCRIPTION
**Pull request by cedricherzog-passbolt**
# Add Caching System

This commit implements a comprehensive caching system for the Passbolt Go SDK, significantly improving performance while maintaining security through proper memory management and secure key zeroing.

## Key Features

**Performance Improvements**
- **Session key caching**: 2x-50x faster metadata decryption by caching session keys per metadata key ID
- **Resource types caching**: Eliminates redundant API calls for resource type lookups
- **Metadata keys caching**: Stores both encrypted and decrypted metadata keys for instant access

**Security Enhancements**
- All cryptographic keys are securely zeroed before being cleared from memory
- Uses ProtonMail gopenpgp v3's `ClearPrivateParams()` for private keys
- Session keys are cloned before caching to prevent corruption
- Enhanced `Logout()` securely wipes all cached keys from memory

## Breaking Changes

**Client becomes permanently invalid after logout**:
- After calling `Logout()`, the client cannot be reused or re-logged in
- User private key is securely zeroed and cleared
- All cached keys are wiped from memory
- For a new session, create a new client instance with `NewClient()`

## New Public API

```go
// Cache management
ClearCache(), ClearResourceTypesCache(), ClearMetadataKeysCache(), ClearSessionKeyCache()

// Cached getters
GetResourceTypesCached(), GetResourceTypeCached(), GetResourceTypeBySlugCached()
GetMetadataKeysCached(), GetDecryptedMetadataKeyCached()

// Session key cache
GetSessionKey(), SetSessionKey()
DecryptMetadataWithKeyID() // Enhanced metadata decryption with caching
```

## Testing

- **1,568 lines of new tests** across 3 test files
- Tests cover secure zeroing, cache behavior, performance, and edge cases

## Statistics

- **8 files changed**: 1,884 insertions, 21 deletions
- Performance: 2x-50x improvement for repeated metadata decryptions

### CODE REVIEW

**Comment by speatzle** [comment]
I think this can be merged as soon as the Session decryption works properly.

File: api/client.go [hunk 1]
```diff
@@ -277,3 +315,311 @@ func (c *Client) GetPGPHandle() *crypto.PGPHandle {
...
+	defer c.sessionKeyCacheMu.Unlock()
+	c.sessionKeyCache[sessionKeyCachePrefixResource+resourceID] = sessionKey
+}
+
+// GetSessionKeyByMetadataKeyID retrieves a cached session key by metadata key ID
+// These session keys are extracted during decrypt and cached as fallback
+func (c *Client) GetSessionKeyByMetadataKeyID(metadataKeyID string) *crypto.SessionKey {
+	c.sessionKeyCacheMu.RLock()
+	defer c.sessionKeyCacheMu.RUnlock()
+	return c.sessionKeyCache[sessionKeyCachePrefixMetaKey+metadataKeyID]
```
Comment (speatzle) [inline api/client.go]: Same as above

File: api/client.go [hunk 2]
```diff
@@ -277,3 +315,311 @@ func (c *Client) GetPGPHandle() *crypto.PGPHandle {
...
+
+	return sessionCount, metadataKeyCount, nil
+}
+
+// GetSessionKey retrieves a cached session key for a metadata key ID (legacy method)
+// For new code, prefer GetSessionKeyByResourceID or GetSessionKeyByMetadataKeyID
+func (c *Client) GetSessionKey(metadataKeyID string) *crypto.SessionKey {
+	c.sessionKeyCacheMu.RLock()
+	defer c.sessionKeyCacheMu.RUnlock()
+	return c.sessionKeyCache[metadataKeyID]
```
Comment (speatzle) [inline api/client.go]: I still think we should return a Copy here instead of a pointer to prevent the API Consumer from accidentally Clearing a Cached key.

File: api/client.go [hunk 3]
```diff
@@ -277,3 +315,311 @@ func (c *Client) GetPGPHandle() *crypto.PGPHandle {
...
+
+	metadataPrivateKeyObj, err := GetPrivateKeyFromArmor(data.ArmoredKey, []byte(data.Passphrase))
+	if err != nil {
+		return nil, fmt.Errorf("Get Metadata Private Key: %w", err)
+	}
+
+	// Cache the decrypted key
+	c.decryptedMetadataKeysCache[id] = metadataPrivateKeyObj
+
+	return metadataPrivateKeyObj, nil
```
Comment (speatzle) [inline api/client.go]: This should also be a copy to prevent the Cached Key from Getting Cleared by the API Consumer.

File: api/client.go [hunk 4]
```diff
@@ -277,3 +315,311 @@ func (c *Client) GetPGPHandle() *crypto.PGPHandle {
...
+		// Log but don't fail - this is also optional optimization
+		c.log("Warning: Failed to pre-decrypt metadata keys: %v", err)
+		err = nil
+	}
+
+	return sessionCount, metadataKeyCount, nil
+}
+
+// GetSessionKey retrieves a cached session key for a metadata key ID (legacy method)
+// For new code, prefer GetSessionKeyByResourceID or GetSessionKeyByMetadataKeyID
```
Comment (speatzle) [inline api/client.go]: Do we even want to add this function if it is already Obsolete?

File: api/client.go [hunk 5]
```diff
@@ -277,3 +315,311 @@ func (c *Client) GetPGPHandle() *crypto.PGPHandle {
...
+func (c *Client) GetSessionKey(metadataKeyID string) *crypto.SessionKey {
+	c.sessionKeyCacheMu.RLock()
+	defer c.sessionKeyCacheMu.RUnlock()
+	return c.sessionKeyCache[metadataKeyID]
+}
+
+// SetSessionKey stores a session key in the cache for a metadata key ID (legacy method)
+// For new code, prefer SetSessionKeyByResourceID or SetSessionKeyByMetadataKeyID
```
Comment (speatzle) [inline api/client.go]: Also obsolete?

**Review by speatzle — CHANGES_REQUESTED** [review]

**Review by speatzle — APPROVED** [review]
